hedera-state-validator/src/main/java/com/hedera/statevalidation/validators/state/Rehash.java

```diff
@@ -81,4 +101,123 @@ private String filterLines(List<String> lines) {
         }
         return "root hash not found";
     }
+
+    @SuppressWarnings("rawtypes")
+    public Hash rehashVm(@NonNull final VirtualMap virtualMap) {
+        final int MAX_FULL_REHASHING_TIMEOUT = 3600; // 1 hour
+        final int MAX_REHASHING_BUFFER_SIZE = 10_000_000; // copied from VirtualMap class
+        final VirtualMapConfig virtualMapConfig = InitUtils.getConfiguration().getConfigData(VirtualMapConfig.class);
+        final VirtualMapStatistics statistics = new VirtualMapStatistics(virtualMap.getLabel());
+        final VirtualDataSource dataSource = virtualMap.getDataSource();
+        final RecordAccessor records = virtualMap.getRecords();
+        requireNonNull(records, "Records must be initialized before rehashing");
+
+        final ConcurrentBlockingIterator<VirtualLeafBytes> rehashIterator =
+                new ConcurrentBlockingIterator<>(MAX_REHASHING_BUFFER_SIZE);
+        final CompletableFuture<Hash> fullRehashFuture = new CompletableFuture<>();
+        final CompletableFuture<Void> leafFeedFuture = new CompletableFuture<>();
+        // getting a range that is relevant for the data source
```

Comment:
The range should come from the virtual map, not from the data source. In this particular case the ranges may match, but in general this is not true